File: example/java/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/example/cache/RedisHashBinder.java
package org.babyfish.jimmer.sql.example.cache;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.babyfish.jimmer.meta.ImmutableProp;
import org.babyfish.jimmer.meta.ImmutableType;
import org.babyfish.jimmer.sql.cache.spi.AbstractRemoteHashBinder;
import org.springframework.dao.DataAccessException;
import org.springframework.data.redis.core.RedisOperations;
import org.springframework.data.redis.core.SessionCallback;

import java.time.Duration;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class RedisHashBinder<K, V> extends AbstractRemoteHashBinder<K, V> {

    private final RedisOperations<String, byte[]> operations;

    public RedisHashBinder(
            RedisOperations<String, byte[]> operations,
            ObjectMapper objectMapper,
            ImmutableType type,
            Duration duration
    ) {
        super(objectMapper, type, null, duration, 30);
        this.operations = operations;
    }

    public RedisHashBinder(
            RedisOperations<String, byte[]> operations,
            ObjectMapper objectMapper,
            ImmutableProp prop,
            Duration duration
    ) {
        super(objectMapper, null, prop, duration, 30);
        this.operations = operations;
    }

    @SuppressWarnings("unchecked")
    @Override
    protected List<byte[]> read(Collection<String> keys, String hashKey) {
        return (List<byte[]>)(List<?>)operations.executePipelined(
                new SessionCallback<Void>() {
                    @Override
                    public <XK, XV> Void execute(RedisOperations<XK, XV> pops) throws DataAccessException {
                        RedisOperations<String, byte[]> pipelinedOps = (RedisOperations<String, byte[]>)pops;
                        for (String key : keys) {
                            pipelinedOps.opsForHash().get(key, hashKey);
                        }
                        return null;
                    }
                }
        );
    }

    @SuppressWarnings("unchecked")
    @Override
    protected void write(Map<String, byte[]> map, String hashKey) {
        operations.executePipelined(
                new SessionCallback<Void>() {
                    @Override
                    public <XK, XV> Void execute(RedisOperations<XK, XV> pops) throws DataAccessException {
                        RedisOperations<String, byte[]> pipelinedOps = (RedisOperations<String, byte[]>)pops;
                        for (Map.Entry<String, byte[]> e : map.entrySet()) {
                            pipelinedOps.opsForHash().put(e.getKey(), hashKey, e.getValue());
                            pipelinedOps.expire(
                                    e.getKey(),
                                    randomMillis(),
                                    TimeUnit.MILLISECONDS
                            );
                        }
                        return null;
                    }
                }
        );
    }

    @Override
    protected void delete(Collection<String> keys) {
        operations.delete(keys);
    }

    @Override
    protected String reason() {
        return "redis";
    }
}

File: example/kotlin/jimmer-sql-graphql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/graphql/cache/binder/RedisHashBinder.kt
package org.babyfish.jimmer.example.kt.graphql.cache.binder

import com.fasterxml.jackson.databind.ObjectMapper
import org.babyfish.jimmer.meta.ImmutableProp
import org.babyfish.jimmer.sql.kt.cache.spi.AbstractKRemoteHashBinder
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisOperations
import org.springframework.data.redis.core.SessionCallback
import java.time.Duration
import java.util.concurrent.TimeUnit

class RedisHashBinder<K, V>(
    private val operations: RedisOperations<String, ByteArray>,
    objectMapper: ObjectMapper?,
    prop: ImmutableProp,
    duration: Duration
) : AbstractKRemoteHashBinder<K, V>(
    objectMapper,
    null,
    prop,
    duration,
    30
) {
    @Suppress("UNCHECKED_CAST")
    override fun read(keys: Collection<String>, hashKey: String): List<ByteArray?> =
        operations.executePipelined(
            object : SessionCallback<Void?> {
                override fun <XK, XV> execute(pops: RedisOperations<XK, XV>): Void? {
                    val pipelinedOps = pops as RedisOperations<String, ByteArray>
                    val hashOps = pipelinedOps.opsForHash<String, ByteArray>()
                    for (key in keys) {
                        hashOps.get(key, hashKey)
                    }
                    return null
                }
            }
        ) as List<ByteArray?>

    @Suppress("UNCHECKED_CAST")
    override fun write(map: Map<String, ByteArray>, hashKey: String) {
        operations.executePipelined(
            object : SessionCallback<Void?> {
                override fun <XK, XV> execute(pops: RedisOperations<XK, XV>): Void? {
                    val pipelinedOps = pops as RedisOperations<String, ByteArray>
                    val hashOps = pipelinedOps.opsForHash<String, ByteArray>()
                    for ((key, value) in map) {
                        hashOps.put(key, hashKey, value)
                        pipelinedOps.expire(
                            key,
                            nextExpireMillis(),
                            TimeUnit.MILLISECONDS
                        )
                    }
                    return null
                }
            }
        )
    }

    override fun delete(keys: Collection<String>) {
        LOGGER.info("Delete property data from redis: {}", keys)
        operations.delete(keys)
    }

    override fun reason(): String = "redis"

    companion object {
        private val LOGGER = LoggerFactory.getLogger(RedisHashBinder::class.java)
    }
}